app/main.py

```diff
@@ -1,6 +1,23 @@
-from typing import Callable
+from typing import Callable, Any
+from functools import wraps
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cache_storage = {}
+
+    @wraps(func)
+    def inner(*args, **kwargs) -> Any:
+        # Checking for Immutability
+        key = func.__name__, args, tuple(kwargs.items())
+        if any(isinstance(arg, (list, dict, set)) for arg in args) or \
+           any(isinstance(arg, (list, dict, set)) for arg in kwargs.values()):
+            raise TypeError("All arguments must be immutable")
```

Comment:
```suggestion
        key = func.__name__, args, tuple(kwargs.items())
```
this check redundant here